### PR TITLE
2.5 - Ensure Synchronous KVM Image Acquisition

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch ensures that parallel provisioning of KVM machines on a single host does not attempt to sync the same backing image to disk multiple times concurrently.

The change involves separating the sync logic from the `Start()` method and calling this separately in the manager, protected by a lock.

The implementation is not ideal - a more elegant solution was abandoned due to the upheaval it caused in the test suite. Ideally this manager would undergo a significant re-write, but we defer to pragmatism.

## QA steps

- Bootstrap to where hosts support KVMs (I used GUIMAAS).
- `juju add-machine`
- `juju deploy redis -n 5 --to kvm:0,kvm:0,kvm:0,kvm:0,kvm:0`
- `watch -c juju status --color`
Observe that only one of the pending machines downloads the image, then all others come up afterwards.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1816170

### CODE REVIEW

File: container/kvm/container.go [hunk 1]
```diff
@@ -35,30 +35,32 @@ func (c *kvmContainer) Name() string {
...
+// the input start parameters has been cached on disk.
+func (c *kvmContainer) EnsureCachedImage(params StartParams) error {
 	var srcFunc func() simplestreams.DataSource
 	if params.ImageDownloadURL != "" {
 		srcFunc = func() simplestreams.DataSource {
 			return imagedownloads.NewDataSource(params.ImageDownloadURL)
 		}
 	}
-	var ftype = BIOSFType
+	var fType = BIOSFType
```
Comment (manadart) [inline container/kvm/container.go]: Sorry about the noise from these - just drive-by appeasements of the spell checker, and the linter complaining about import name collisions.

**Comment by manadart** [comment]
[KVM instances fail at creation with: Failed to get shared "write" lock Is another process using the image?](https://trello.com/c/uaRJUYaW/1188-kvm-instances-fail-at-creation-with-failed-to-get-shared-write-lock-is-another-process-using-the-image)

**Comment by dshcherb** [comment]
Did some manual QA on this change:

```
ubuntu@maas:~$ juju status
Model    Controller  Cloud/Region  Version  SLA          Timestamp
kvmtest  samaas      samaas        2.5.3    unsupported  19:18:51Z

App     Version  Status  Scale  Charm   Store       Rev  OS      Notes
ubuntu  18.04    active      6  ubuntu  jujucharms   12  ubuntu  

Unit        Workload  Agent  Machine  Public address  Ports  Message
ubuntu/31   active    idle   4/kvm/0  192.0.2.22            ready
ubuntu/32   active    idle   4/kvm/1  192.0.2.248           ready
ubuntu/33   active    idle   4/kvm/2  192.0.2.88            ready
ubuntu/34*  active    idle   4/kvm/3  192.0.2.23            ready
ubuntu/35   active    idle   4/kvm/4  192.0.2.70            ready
ubuntu/36   active    idle   4/kvm/5  192.0.2.84            ready

Machine  State    DNS           Inst id              Series  AZ       Message
4        started  192.0.2.16  empusa               bionic  default  Deployed
4/kvm/0  started  192.0.2.22   juju-487449-4-kvm-0  bionic           Container started
4/kvm/1  started  192.0.2.248  juju-487449-4-kvm-1  bionic           Container started
4/kvm/2  started  192.0.2.88   juju-487449-4-kvm-2  bionic           Container started
4/kvm/3  started  192.0.2.23   juju-487449-4-kvm-3  bionic           Container started
4/kvm/4  started  192.0.2.70   juju-487449-4-kvm-4  bionic           Container started
4/kvm/5  started  192.0.2.84   juju-487449-4-kvm-5  bionic           Container started
```

Looks good to me.

**Comment by mitechie** [comment]
Thank you for the QA @dshcherb. We're code reviewing the branch currently and should have it in the edge snap soon. From there we'll wrap up 2.5.3 and try to get a candidate put together quickly.

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by manadart** [comment]
@hmlanigan Thanks for shepherding this one in. I'll think about some suitable tests for the logic.

File: container/kvm/kvm.go [hunk 2]
```diff
@@ -213,14 +215,25 @@ func (manager *containerManager) CreateContainer(
...
-	callback(status.Provisioning, "Creating container; it might take some time", nil)
+	_ = callback(status.Provisioning, "Creating container; it might take some time", nil)
 	logger.Tracef("create the container, constraints: %v", cons)
+
+	// Lock around finding an image.
+	// The provisioner works concurrently to create containers.
+	// If an image needs to be copied from a remote, we don't many goroutines
+	// attempting to do it at once.
+	manager.imageMutex.Lock()
+	err = kvmContainer.EnsureCachedImage(startParams)
```
Comment (cpaelzer) [inline container/kvm/kvm.go]: shouldn't that lock better be "inside" the function to ensure if it is ever called from another place that they still use the same lock? If you prefer it here then the function might want a way to check if the right lock is held maybe?

File: container/kvm/container.go (hunk 1 above)
Comment (cpaelzer) [reply container/kvm/container.go]: Why don't you split the
- fType renames
- the _ = assignment changes
- shortening of container to c
- the logDir renames

Into one commit each?
That would make the actual change that is meant to be done (synchronize image activities) much more reviewable.

File: container/kvm/kvm.go (hunk 2 above)
Comment (manadart) [reply container/kvm/kvm.go]: The mutex itself needs to be a member of the (singular) manager rather than the container object, as that's the level at which we have to gate access.

We could either instantiate the container with the mutex passed in, or call the method passing the mutex, but handing around mutex references has a smell about it.

This code has parts that are half a decade old. It's not ideal but I've gone for the "lightest touch". It has also never been Goroutine safe in that time - I have a follow-up patch that will address this.

File: container/kvm/container.go (hunk 1 above)
Comment (manadart) [reply container/kvm/container.go]: Thanks; that is generally my approach.

**Comment by dshcherb** [comment]
Haven't tried the follow-up patch yet but with this one I managed to get into an interesting state (data store ISOs for different VMs contain similar data - notice how hostnames are the same):

```
ubuntu@control-8:~$ virsh console juju-d703a2-6-kvm-0
Connected to domain juju-d703a2-6-kvm-0
Escape character is ^]

Ubuntu 18.04.2 LTS juju-d703a2-6-kvm-0 ttyS0

juju-d703a2-6-kvm-0 login:
ubuntu@control-8:~$ virsh console juju-d703a2-6-kvm-1
Connected to domain juju-d703a2-6-kvm-1
Escape character is ^]

Ubuntu 18.04.2 LTS juju-d703a2-6-kvm-0 ttyS0

juju-d703a2-6-kvm-0 login:

root@control-8:~# ls -al  /var/lib/juju/kvm/guests/
total 4319188
drwxr-xr-x 2 libvirt-qemu kvm        4096 Mar 19 11:54 .
drwxr-xr-x 3 root         root       4096 Mar 19 11:50 ..
-rw-r--r-- 1 libvirt-qemu kvm  2361393152 Mar 19 11:54 bionic-amd64-backing-file.qcow
-rw-r--r-- 1 libvirt-qemu kvm      391168 Mar 19 11:54 juju-d703a2-6-kvm-0-ds.iso
-rw-r--r-- 1 libvirt-qemu kvm  2622750720 Mar 19 12:33 juju-d703a2-6-kvm-0.qcow
-rw-r--r-- 1 libvirt-qemu kvm      391168 Mar 19 11:54 juju-d703a2-6-kvm-1-ds.iso
-rw-r--r-- 1 libvirt-qemu kvm   716570624 Mar 19 12:33 juju-d703a2-6-kvm-1.qcow


pgrep -af qemu | grep -ioP 'drive file=.*?\s'
drive file=/var/lib/juju/kvm/guests/juju-d703a2-6-kvm-1.qcow,format=qcow2,if=none,id=drive-virtio-disk0 
drive file=/var/lib/juju/kvm/guests/juju-d703a2-6-kvm-1-ds.iso,format=raw,if=none,id=drive-virtio-disk1 
drive file=/var/lib/juju/kvm/guests/juju-d703a2-6-kvm-0.qcow,format=qcow2,if=none,id=drive-virtio-disk0 
drive file=/var/lib/juju/kvm/guests/juju-d703a2-6-kvm-0-ds.iso,format=raw,if=none,id=drive-virtio-disk1 

mkdir /mnt/ds-0
mkdir /mnt/ds-1

root@control-8:~# mount -o ro,loop /var/lib/juju/kvm/guests/juju-d703a2-6-kvm-0-ds.iso /mnt/ds-0
root@control-8:~# mount -o ro,loop /var/lib/juju/kvm/guests/juju-d703a2-6-kvm-1-ds.iso /mnt/ds-1
root@control-8:~# grep hostname /mnt/ds-0/user-data
hostname: juju-d703a2-6-kvm-0
root@control-8:~# grep hostname /mnt/ds-1/user-data
hostname: juju-d703a2-6-kvm-0

root@control-8:~# md5sum /var/lib/juju/kvm/guests/juju-d703a2-6-kvm-0-ds.iso
09c10e1cf3a003bab359bf042f231d52  /var/lib/juju/kvm/guests/juju-d703a2-6-kvm-0-ds.iso
root@control-8:~# md5sum /var/lib/juju/kvm/guests/juju-d703a2-6-kvm-1-ds.iso
8a930aab5fcab68c7cd743c901589e57  /var/lib/juju/kvm/guests/juju-d703a2-6-kvm-1-ds.iso

root@control-8:~# md5sum /mnt/ds-0/user-data
c5b20e7458edaaeeeb0c50e3ea8e2a14  /mnt/ds-0/user-data
root@control-8:~# md5sum /mnt/ds-1/user-data
68a39e10e711a70ffd5d46b08800a95d  /mnt/ds-1/user-data

```

This looks like a different problem which I also described here (with data store ISOs):
https://bugs.launchpad.net/juju/+bug/1816170/comments/10
